### PR TITLE
HDanywhere : Add support for new line of matrices + readme.MD

### PR DESCRIPTION
**Pull request by kgoderis**
Signed-off-by: Karel Goderis <karel.goderis@me.com>

### CODE REVIEW

File: addons/binding/org.openhab.binding.hdanywhere/README.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
+# HDanywhere Binding 
+
+This binding integrates with [HDanywhere](http://www.hdanywhere.co.uk) HDMI matrices. 
+
+HDanywhereis a manufacturer of multiroom/distributed audio/video equipment. This binding supports their V3 of the Multiroom+ HDMI matrix running firmware V1.2(20131222), as well as the newer MHUB series of matrices. These matrices support the highest HD resolutions, including 1080p 3D & 4K, use a single Cat5e/6/7 wiring structure with reliable performance up to 100m, have IR passback to allow you to select and control what you watch from every room and are fully compatible with universal remote controls
```
Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/README.md]: HDanywhere is

File: addons/binding/org.openhab.binding.hdanywhere/README.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+The matrixes can be controlled by either UDP/IP and/or Serial connections, but due to the lack of feedback on the actual state of the HDMI matrix when using those methods, this binding operates by controlling the built-in webserver of the matrix.
```
Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/README.md]: matrices

File: addons/binding/org.openhab.binding.hdanywhere/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+## Supported Things
+
+This binding currently supports the following thing types:
+
+* *multiroomplus* : Multiroom+ V3 (**Note:** This product is no longer sold by HDanywhere)
+* *mhub4k431* : MHUB 4K (4X3+1)
+
+## Discovery
+
+The binding does not support a discovery feature
```
Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/README.md]: feature.

File: addons/binding/org.openhab.binding.hdanywhere/README.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+## Binding Configuration
+
+This binding does not require any special configuration.
+
+## Thing Configuration
+
+Each thing requires the IP address of the matrix, and the interval in between status updates that are fetched from the matrix. Additionally, the *multiroomplus* has an additional required parameter 'ports' to specify the number of physical ports (e.g. 4x4, 8x8,...) of the matrix
```
Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/README.md]: matrix.

File: addons/binding/org.openhab.binding.hdanywhere/README.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+| Thing Type      |  Item Type    | Description                                                         |
+|-----------------|---------------|---------------------------------------------------------------------|
+| port1           |  Number       | The number of the input port that is connected to the output port 1 |
+| port2           |  Number       | The number of the input port that is connected to the output port 2 |
+| port3           |  Number       | The number of the input port that is connected to the output port 3 |
+| ...             |  Number       | ...                                                                 |
+| port8           |  Number       | The number of the input port that is connected to the output port 8 |
+
+
+## Binding Configuration
```
Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/README.md]: Duplicate of existing section above.

File: addons/binding/org.openhab.binding.hdanywhere/OSGI-INF/HDanywhereHandlerFactory.xml [hunk 6]
```diff
@@ -10,11 +10,13 @@
 
 -->
 <scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="org.openhab.binding.hdanywhere.internal.HDanywhereHandlerFactory">
-
+  
    <implementation class="org.openhab.binding.hdanywhere.internal.HDanywhereHandlerFactory"/>
-
+   <reference bind="setItemChannelLinkRegistry" cardinality="1..1" interface="org.eclipse.smarthome.core.thing.link.ItemChannelLinkRegistry" name="ItemChannelLinkRegistry" policy="static" unbind="unsetItemChannelLinkRegistry"/>
```
Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/OSGI-INF/HDanywhereHandlerFactory.xml]: So bindings are allowed to use the ItemChannelLinkRegistry now?

Comment (kgoderis) [inline addons/binding/org.openhab.binding.hdanywhere/OSGI-INF/HDanywhereHandlerFactory.xml]: @watou ;-) I see no reason why they would not be allowed to do so, do you?

Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/OSGI-INF/HDanywhereHandlerFactory.xml]: I can't say that I know enough to have a fully informed opinion, but with that huge caveat, no I don't know of a reason! ;-)

Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.hdanywhere/OSGI-INF/HDanywhereHandlerFactory.xml]: No, it is not, see [here](https://www.eclipse.org/smarthome/documentation/development/bindings/dependencies.html).
I guess Karel wants to work around the fact that https://github.com/eclipse/smarthome/issues/595 isn't yet implemented.

Comment (watou) [inline addons/binding/org.openhab.binding.hdanywhere/OSGI-INF/HDanywhereHandlerFactory.xml]: Same reason it's now used in KNX 2.0 binding?

**Comment by kgoderis** [comment]
here the idea is to overrule whatever is set by means of the AutoUpdate binding, eg the binding will report back the actual state of a Channel, and does not need the runtime to update the Item

**Comment by kgoderis** [comment]
The list of bundles reads as list of suggestive bundles to use, and not as an exhaustive list, at least to me it does

@kaikreuzer Shall we proceed with the workaround until the disable feature is provided?

**Comment by kaikreuzer** [comment]
> @kaikreuzer Shall we proceed with the workaround until the disable feature is provided?

No, I don't think so. What is the issue with having the state update event? Your binding can still send the "final" state itself as soon as it can provide it? That's how it works for almost all bindings.

**Comment by kgoderis** [comment]
The fact that you have 2 updates; and when it does not update it would indicate that there is some kind of problem with the device, instead of going unnoticed.

**Comment by kaikreuzer** [comment]
That's absolutely fine and just the way it is designed. Consider the first update to be the "optimistic, but quick guess" and the second as the "definitive, but slow update".
Problems with devices should be anyhow identified by the binding and reported through the Thing status.

**Comment by kgoderis** [comment]
@kaikreuzer done with the small fixes + auto-update revert.

File: addons/binding/org.openhab.binding.hdanywhere/src/main/java/org/openhab/binding/hdanywhere/handler/Mhub4K431Handler.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                InputStream stream = new ByteArrayInputStream(content.getBytes(StandardCharsets.UTF_8));
+
+                Properties httpHeaders = new Properties();
+                httpHeaders.setProperty("Cookie", "logintype-88=01");
+
+                try {
+                    String response = HttpUtil.executeUrl(httpMethod, url, httpHeaders, stream,
+                            "application/x-www-form-urlencoded; charset=UTF-8", timeout);
+                } catch (IOException e) {
+                    logger.error("Communication with device failed: {}", e);
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.hdanywhere/src/main/java/org/openhab/binding/hdanywhere/handler/Mhub4K431Handler.java]: could you set the ThingStatus to OFFLINE with an according message and reduce the logging to debug?

File: addons/binding/org.openhab.binding.hdanywhere/META-INF/MANIFEST.MF [hunk 8]
```diff
@@ -7,11 +7,15 @@ Bundle-Version: 2.1.0.qualifier
 Bundle-RequiredExecutionEnvironment: JavaSE-1.7
 Bundle-ClassPath: .
 Import-Package: com.google.common.collect,
+ com.google.gson,
+ com.google.gson.reflect,
  org.apache.commons.lang,
  org.eclipse.smarthome.config.core,
+ org.eclipse.smarthome.core.autoupdate,
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.hdanywhere/META-INF/MANIFEST.MF]: I guess this line can be removed again

File: addons/binding/org.openhab.binding.hdanywhere/META-INF/MANIFEST.MF [hunk 9]
```diff
@@ -7,11 +7,15 @@ Bundle-Version: 2.1.0.qualifier
...
  org.eclipse.smarthome.core.library.types,
  org.eclipse.smarthome.core.thing,
  org.eclipse.smarthome.core.thing.binding,
+ org.eclipse.smarthome.core.thing.link,
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.hdanywhere/META-INF/MANIFEST.MF]: dito

**Comment by kgoderis** [comment]
@kaikreuzer done.

**Pull request merged**